[FE] refactor: 바텀시트가 사용자의 움직임보다 느리게 동작하는 문제 개선 및 성능 최적화

kaori-killer opened the pull request:
# #️⃣ Issue Number

#505 

## 🕹️ 작업 내용

- 배경/문제: 바텀시트를 손가락으로 위아래로 드래그할 때, 모바일 프로덕션 환경에서만 움직임이 느리고 둔하게 느껴지는 문제가 발생했습니다. 개발 환경에서는 정상 동작했기에 원인 파악이 어려웠습니다. DevTools로 분석해보니 드래그할 때마다 CPU 사용률이 92.3%까지 치솟고 있었습니다.
- 목표/의도: CPU에게 과도한 일을 시키지 않도록 개선하여, 사용자 손가락 움직임을 자연스럽게 따라오는 60FPS 애니메이션을 구현했습니다.
- 작업 내용 요약
  - [x] height 상태 관리를 React State에서 useRef 기반 DOM 직접 조작으로 전환 
  - [x] CSS transform 속성 적용으로 하드웨어 가속 활성화
  - [x] will-change, transform3D 속성으로 레이어 사전 분리
  - [x] React.memo 적용으로 불필요한 리렌더링 차단

## 🎨 Before / After (Optional)

| 항목 | Before | After |
|------|--------|-------|
| **애니메이션** | [실행영상](https://github.com/user-attachments/assets/8b839518-52ea-4fb4-af71-d548d6fff47f) | [실행영상](https://github.com/user-attachments/assets/d6e2bf57-39f9-4bd0-931e-c3755a78ab94) |
| **퍼포먼스 탭** | <img width="958" height="261" alt="스크린샷 2025-11-16 오전 3 02 27" src="https://github.com/user-attachments/assets/141ed4b0-38e4-4eb2-b933-639bef415153" /> | <img width="958" height="266" alt="스크린샷 2025-11-16 오전 3 02 09" src="https://github.com/user-attachments/assets/da169a01-0e75-42d8-a153-fa48cbf542ff" /> |
| **CPU 사용률** | 최대 92%, 불규칙적 프레임 드랍 | 최대 64%, 안정적인 60FPS 유지  |
| **렌더링 파이프라인** | 스타일 재계산 → CSSOM 업데이트 → 레이아웃 재계산 반복 | Composite 단계만 실행 |
| **상태 관리** | React State로 height 관리 (리렌더링 발생) | useRef로 DOM 직접 조작 (리렌더링 없음) |

## 📋 리뷰 포인트

- [ ] 모바일 프로덕션 환경에서  바텀시트 드래그 시 부드럽게 동작하는지 확인해주세요.
- [ ] 바텀시트를 제외한 다른 기능이 여전히 제대로 동작하는지 확인해주세요.

##  💬 Discussion (깊은 의견 교환용)

바텀시트의 height 관리를 React State 대신 useRef를 통한 명령형 DOM 조작으로 다뤘습니다. [React 공식문서: Ref로 DOM 조작하기](https://ko.react.dev/learn/manipulating-the-dom-with-refs)에서도 이러한 사용 사례를 인정하는 것 같습니다. 성능을 위해 DOM 조작을 시도해봤는데 이 구현에 대해 어떻게 생각하시나요?

## 🔮 더 자세한 구현 과정

구현 과정을 [블로그](https://velog.io/@kaori-killer/%EB%AA%A8%EB%B0%94%EC%9D%BC-%EB%B0%94%ED%85%80%EC%8B%9C%ED%8A%B8-%EB%93%9C%EB%9E%98%EA%B7%B8-%EC%8B%9C-%EC%84%B1%EB%8A%A5-%EC%A0%80%ED%95%98%EA%B0%80-%EB%B0%9C%EC%83%9D%ED%95%98%EB%8A%94-%ED%98%84%EC%83%81-%EB%94%94%EB%B2%84%EA%B9%85%ED%95%98%EA%B8%B0)로 작성했습니다. 리뷰하실 때 참고부탁드려요!

## 🤔 더 공부해야 하는 부분

- [ ] 브라우저의 레이어 생성 기준은 무엇일까?
- [ ] will-change 속성이 과도한 레이어 생성으로 메모리 이슈를 일으키지는 않을까?

kaori-killer committed:
refactor: 바텀시트의 높이를 transition에 height에서 transition로 게산하도록 변경

kaori-killer committed:
refactor:  성능 향상을 위해 bottomSheetView의  transform 속성을 translate3d로 업데이트합니다.

kaori-killer committed:
refactor: 바텀시트 내부에 불필요하게 리렌더링 되는 부분 React.memo 처리

kaori-killer committed:
refactor: 바텀시트 드래그 시 React 상태 대신 DOM 직접 조작으로 프레임 드랍 해결

kaori-killer committed:
chore: HTMLDivElement로 타입 설정할 때 > 빠진 오타 수정

kaori-killer committed:
fix: React 상태 동기화, 애니메이션이 시작된 후에 인라인 스타일을 제거하도록 수정

- React 상태 동기화, 애니메이션이 시작된 후에 인라인 스타일을 제거함

eunsoA commented:
> 구현 과정을 [블로그](https://velog.io/@kaori-killer/%EB%AA%A8%EB%B0%94%EC%9D%BC-%EB%B0%94%ED%85%80%EC%8B%9C%ED%8A%B8-%EB%93%9C%EB%9E%98%EA%B7%B8-%EC%8B%9C-%EC%84%B1%EB%8A%A5-%EC%A0%80%ED%95%98%EA%B0%80-%EB%B0%9C%EC%83%9D%ED%95%98%EB%8A%94-%ED%98%84%EC%83%81-%EB%94%94%EB%B2%84%EA%B9%85%ED%95%98%EA%B8%B0)로 작성했습니다. 리뷰하실 때 참고부탁드려요!

블로그 글이 없다고 뜨는데요 ~ 확인해주실수있나요 ~?

kaori-killer commented:
> 블로그 글이 없다고 뜨는데요 ~ 확인해주실수있나요 ~?

블로그 글의 제목을 바꿨더니 URL이 변경됐네요!
[모바일 환경에서 바텀시트 드래그 시 성능 저하가 발생하는 현상 디버깅하기](https://velog.io/@kaori-killer/%EB%AA%A8%EB%B0%94%EC%9D%BC-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C-%EB%B0%94%ED%85%80%EC%8B%9C%ED%8A%B8-%EB%93%9C%EB%9E%98%EA%B7%B8-%EC%8B%9C-%EC%84%B1%EB%8A%A5-%EC%A0%80%ED%95%98%EA%B0%80-%EB%B0%9C%EC%83%9D%ED%95%98%EB%8A%94-%ED%98%84%EC%83%81-%EB%94%94%EB%B2%84%EA%B9%85%ED%95%98%EA%B8%B0) 여기 있습니다 😊

kaori-killer committed:
fix: CSS 내 transform 제거하여 transform의 중복 선언으로 인한 충돌 방지

kaori-killer committed:
fix: positionPercent 상태가 바뀔 때마다 바텀시트의 실제 위치(transform)를 업데이트

- 기존에는 드래그 끝나면 inline style에 있는  transform을 지우고, styled CSS에 있는 transform이 다시 적용되면서 CSS transition이 발동되어 애니메이션과 함께 실제 위치 업데이트가 가능했음

kaori-killer commented:
# 바텀시트 뒤에 있는 지도 클릭 안 되는 버그 수정

## 문제 상황

바텀시트를 아래로 접어도 뒤에 있는 지도를 클릭할 수 없었습니다.

| BEFORE | AFTER |
| ----- | ----- |
| ![wrapper_before_바텀시트](https://github.com/user-attachments/assets/27a9fd4c-b924-44eb-8932-cd364a760162) | ![wrapper_after_바텀시트 gif](https://github.com/user-attachments/assets/26ca6dcb-85a6-41ad-84d3-772d929cdd8e) |
| 뒤에 지도 클릭 안됨 | 뒤에 지도 클릭 됨 |

### 원인

<img width="670" alt="구조 설명" src="https://github.com/user-attachments/assets/e3cafe54-fe40-4e59-b8ce-91f312539e1b" />

바깥 wrapper는 `position: fixed`로 고정되어 있고, 안쪽 container만 `transform`으로 이동하여 wrapper가 계속 지도를 가리고 있었습니다.

```tsx
// Before
<div css={[bottomSheetView.base()]}>  {/* 항상 화면 하단 고정 */}
  <div ref={containerRef}>  {/* 이것만 transform으로 움직임 */}
```

```tsx
// AFTER
<div css={[bottomSheetView.base()]} ref={containerRef}> {/* transform으로 움직임 */}
  <div>
</div>
```
- 관련 커밋: 
  - 50a1f4cf6c8fc5afff4c831afdbb1c89fd99e10b
  - 36836326cb84354cdc7ce51fd47ed7ce7c840b4c

eunsoA commented on frontend/src/features/recommendation/components/bottomSheetView/bottomSheetView.styled.ts:
```diff
@@ -13,29 +12,30 @@ export const base = () => css`
-  height: ${MIN_VH + (MAX_VH - MIN_VH) * (positionPercent / 100)}dvh;
-  min-height: ${MIN_VH}dvh;
-  max-height: ${MAX_VH}dvh;
+  height: ${MAX_VH}dvh;
```
여기서 최소 높이와 `positionPercent`로 높이를 계산하는 로직을 지워주신 의도가 있을까요?

최소 높이가 사라지면서 아래와 같이 모바일 환경에서 바텀시트가 아예 사라져버리는 이슈가 생기고 있어요. (하단의 영상 첨부해요!)


https://github.com/user-attachments/assets/706d6078-03ad-4aee-8e97-aef0831c7762

eunsoA commented on frontend/src/features/recommendation/components/bottomSheetDetail/BottomSheetDetail.tsx:
```diff
@@ -38,4 +40,4 @@ function BottomSheetDetail({
 }
 
-export default BottomSheetDetail;
+export default React.memo(BottomSheetDetail);
```
`BottomSheetDetail`과 `BottomSheetList` 둘다 React.memo로 감싸주신 것 좋네요! 👍👍

eunsoA commented on frontend/src/features/recommendation/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -37,11 +37,18 @@ function BottomSheet({
   const startYRef = useRef(0);
   const startPercentRef = useRef(positionPercent);
   const viewportRef = useRef<number>(getViewportHeight());
+  const currentPercentRef = useRef(positionPercent); // 드래그 중 현재 위치 추적
+  const containerRef = useRef<HTMLDivElement>(null); // DOM 직접 조작용
 
   const activePointerIdRef = useRef<number | null>(null);
 
   useSyncViewportHeight(viewportRef);
 
+  // positionPercent가 변경될 때 currentPercentRef도 동기화
+  useEffect(() => {
+    currentPercentRef.current = positionPercent;
+  }, [positionPercent]);
```
성능 최적화를 위해 `positionPercent`(react state)와 `currentPercentRef`(ref)를 분리해주신 것 같네요!

> 바텀시트의 height 관리를 React State 대신 useRef를 통한 명령형 DOM 조작으로 다뤘습니다. [React 공식문서: Ref로 DOM 조작하기](https://ko.react.dev/learn/manipulating-the-dom-with-refs)에서도 이러한 사용 사례를 인정하는 것 같습니다. 성능을 위해 DOM 조작을 시도해봤는데 이 구현에 대해 어떻게 생각하시나요?

React state가 아닌 Ref를 사용하여 드래그 중 과도한 리렌더링을 방지하는 것은 Ref를 사용하는 적절한 최적화 사례라고 생각합니다! 👍

---

다만 실행해보니 한 가지 이슈가 있는 것 같아요!

### [이슈] 드래그 종료 시 깜빡임 & transition 미작동
**현상:**
- 드래그를 놓으면 스냅 포인트로 애니메이션 없이 즉시 이동 (깜빡임)
- `handleTransitionEnd`가 실행되지 않아 `isAnimating`이 `false`로 돌아가지 않음.

현재 transform을 조작하는 요소와 transition이 적용된 요소가 달라서 생기는 이슈인 것 같은데요 이 부분에 대해서는 어떻게 생각하시나요 ?

https://github.com/user-attachments/assets/7eabb063-8c7f-4937-8db3-9e117a3930c0

---
제가 전체 로직을 다 이해하지 못한 질문일 수도 있기에 편하게 답변주시면 감사하겠습니다 !

eunsoA reviewed (changes requested):
안녕하세요 헤일리~! 🍊

관심있던 내용이라 상당히 재미있게 리뷰했던 것 같습니다!
게다가 정리된 블로그 첨부해주셔서 따라가기 더 쉬웠던 것 같습니다. 깔끔한 설명 감사해요 ~ ☺️

구현해주신 내용 중 몇가지 궁금한 점을 코멘트 남겨놨습니다. 확인 부탁드려요! :)
수고 많으셨습니다!

---

### 📋 리뷰 포인트
- [ ] 모바일 프로덕션 환경에서 바텀시트 드래그 시 부드럽게 동작하는지 확인해주세요.
- [ ] 바텀시트를 제외한 다른 기능이 여전히 제대로 동작하는지 확인해주세요.

---

### [이슈] 투표하기 버튼이 사라졌어요

제목 그대로 투표하기 버튼이 뷰에서 안보이고 있어요. 올려주신 데모 예시에도 안보이는 것 같네요!
확인 부탁드립니다~!

<img width="325" height="675" alt="스크린샷 2025-11-18 오후 3 35 27" src="https://github.com/user-attachments/assets/f93af435-3769-4584-a2ed-20e31c87139a" />

kaori-killer committed:
fix: contain 속성 paint 값 제거로 투표하기 버튼 UI를 화면에 표시

기존에는 contain 속성에 paint 값을 줘서 페인팅을 격리시켰다.
 - 바텀시트 내부만 paint 하도록 격리시켰다.
 - 이는 overflow: hidden과 유사한 시각적 효과를 낸다.
 -  하지만 투표하기 버튼은 바텀시트 밖에 있었다.
 - 이 속성으로 인해 투표하기 UI가 감춰졌다.

kaori-killer commented:
> [이슈] 투표하기 버튼이 사라졌어요
>
> 제목 그대로 투표하기 버튼이 뷰에서 안보이고 있어요. 올려주신 데모 예시에도 안보이는 것 같네요!
확인 부탁드립니다~!


b2ef66123df5208047e39e65b5a14398201a9501
